packages/flutter/lib/src/widgets/spell_check.dart

```diff
@@ -0,0 +1,330 @@
+
+      if (currSpan.range.start > textPointer) {
+        endIndex = currSpan.range.start < text.length
+            ? currSpan.range.start
+            : text.length;
+        textPointerWithinComposingRegion =
+            composingRegion.start >= textPointer &&
+                composingRegion.end <= endIndex &&
+                !composingWithinCurrentTextRange;
+
+        if (textPointerWithinComposingRegion) {
+          _addComposingRegionTextSpans(tsTreeChildren, text, textPointer,
+              composingRegion, style, composingTextStyle);
+          tsTreeChildren.add(
+            TextSpan(
+              style: style,
+              text: text.substring(composingRegion.end, endIndex)
+            )
+          );
+        } else {
+          tsTreeChildren.add(
+            TextSpan(
+              style: style,
+              text: text.substring(textPointer, endIndex)
+            )
+          );
+        }
+
+        textPointer = endIndex;
+      } else {
+        endIndex =
+            currSpan.range.end < text.length ? currSpan.range.end : text.length;
+        currSpanIsComposingRegion = textPointer >= composingRegion.start &&
+            endIndex <= composingRegion.end &&
+            !composingWithinCurrentTextRange;
+        tsTreeChildren.add(
+          TextSpan(
+            style: currSpanIsComposingRegion
+                ? composingTextStyle
+                : misspelledJointStyle,
+            text: text.substring(currSpan.range.start, endIndex)
+          )
+        );
+
+        textPointer = endIndex;
+        currSpanPointer++;
+      }
+    }
+  }
+
+  // Add any remaining text to the tree if applicable.
+  if (textPointer < text.length) {
+    if (textPointer < composingRegion.start &&
+        !composingWithinCurrentTextRange) {
+      _addComposingRegionTextSpans(tsTreeChildren, text, textPointer,
+          composingRegion, style, composingTextStyle);
+
+      if (composingRegion.end != text.length) {
+        tsTreeChildren.add(
+          TextSpan(
+            style: style,
+            text: text.substring(composingRegion.end, text.length)
+          )
+        );
+      }
+    } else {
+      tsTreeChildren.add(
+        TextSpan(
+          style: style, text: text.substring(textPointer, text.length)
+        )
+      );
+    }
+  }
+
+  return tsTreeChildren;
+}
+
+/// Helper method to create [TextSpan] tree children for specified range of
+/// text up to and including the composing region.
+void _addComposingRegionTextSpans(
+    List<TextSpan> treeChildren,
+    String text,
+    int start,
+    TextRange composingRegion,
+    TextStyle? style,
+    TextStyle composingTextStyle) {
+  treeChildren.add(
+    TextSpan(
+      style: style,
+      text: text.substring(start, composingRegion.start)
+    )
+  );
+  treeChildren.add(
+    TextSpan(
+      style: composingTextStyle,
+      text: text.substring(composingRegion.start, composingRegion.end)
+    )
+  );
+}
```

Comment:
I don't see a problem with keeping everything here in its own file in the widgets library :+1: .  Unless @camsim99 or another reviewer has a better idea.

Reply:
Will definitely be useful to group the methods together especially when the menu logic needs to be added somewhere.